### DISCUSSION TITLE
How properly write integration tests with nested transactions

### DISCUSSION THREAD

**Discussion by aramazanov** [sqlalchemy/sqlalchemy#10011] [Usage Questions] [answered]
Hello everyone!

I have seen a lot of answers about that topic but I didn't find the answer.

For example, I have a simple fixture with the db session in my tests:
```
@pytest.fixture(scope="function")
def db_session():
    engine = create_engine()
    connection = engine.connect()

    transaction = connection.begin()
    session = Session(bind=connection)

    yield session

    session.close()
    transaction.rollback()
    connection.close()
```

And I have an integration test:
```
def test_function_with_begin(db_session):
   # Let's call the function with begin()
   function_with_begin()
```

For that function with begin()
```
def function_with_begin():
   # Some atomic actions with commit at the end of transaction
   # Yes, it nested transaction only for tests
   with db.Session.begin():
      ...
  
```

I catch the exception  "A transaction is already begun on this Session." It happens because I call begin() without the parameter nested=True (or just begin_nested()). The first call of begin() was in the db_session fixture.

How can I properly test it without rewriting the code of the application?

I use sqlalchemy==2.0.16 and last postgresql

Thank you in advance!

**Comment by zzzeek** [comment]
the official tested recipe with this is at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites 

this is based on unittest but can be adapted to a pytest fixture directly.

**Reply by aramazanov** [reply]
Hello @zzzeek! Yes, I have seen it. My fixture works the same way except for join_transaction_mode="create_savepoint". But it doesn't help because begin should be called with nested=True. Maybe you have examples for my case?

**Reply by zzzeek** [reply]
`join_transaction_mode="create_savepoint"` is what makes it do that automatically.  use that as documented

**Reply by zzzeek** [reply]
example for your case:


```py
@pytest.fixture(scope="function")
def db_session():
    engine = create_engine()
    # connect to the database
    connection = engine.connect()

    # begin a non-ORM transaction
    trans = connection.begin()

    # bind an individual Session to the connection, selecting
    # "create_savepoint" join_transaction_mode
    session = Session(
            bind=connection, join_transaction_mode="create_savepoint"
        )

    yield session

    session.close()

    # rollback - everything that happened with the
    # Session above (including calls to commit())
    # is rolled back.
    trans.rollback()

    # return connection to the Engine
    connection.close()

    # since you are using create_engine() inside the decorator, dispose it
    engine.dispose()

```

**Reply by aramazanov** [reply]
@zzzeek  thank you for example. But It doesn't work for me, I get the exception at https://github.com/sqlalchemy/sqlalchemy/blob/rel_2_0_16/lib/sqlalchemy/orm/session.py#L1812 . Maybe it has some magick but I don't see special behavior for join_transaction_mode and it looks like nested=True is required anyway.

**Reply by zzzeek** [reply]
your test function "function_with_begin()" seems to refer to some global session yet your pytest fixture refers to a function-local session.  perhaps you are not making use of your fixtured session?  

Here's a demonstration case.  run it as "pytest test.py":

```py
import pytest
from sqlalchemy import create_engine, select

from sqlalchemy.orm import Session


@pytest.fixture(scope="function")
def db_session():
    engine = create_engine("postgresql://scott:tiger@localhost/test")
    # connect to the database
    connection = engine.connect()

    # begin a non-ORM transaction
    trans = connection.begin()

    # bind an individual Session to the connection, selecting
    # "create_savepoint" join_transaction_mode
    session = Session(
            bind=connection, join_transaction_mode="create_savepoint"
        )

    yield session

    session.close()

    # rollback - everything that happened with the
    # Session above (including calls to commit())
    # is rolled back.
    trans.rollback()

    # return connection to the Engine
    connection.close()

    # since you are using create_engine() inside the decorator, dispose it
    engine.dispose()



def test_function_with_begin(db_session):
   # Let's call the function with begin()

   function_with_begin(db_session)


def test_function_with_commit_rollback(db_session):
   # Let's call the function with begin()

   function_with_commit_rollback(db_session)


def function_with_begin(sess):
    with sess.begin():
        sess.execute(select(1))

        sess.execute(select(1))


def function_with_commit_rollback(sess):
    sess.execute(select(1))

    sess.rollback()

    sess.execute(select(1))

    sess.commit()

```

**Reply by aramazanov** [reply]
Hello @zzzeek! I have the same session. And i have found the reason. I'm learning sqlalchemy now but it has strange behavior for me after the Django ORM. It's edited example: 

```python
import pytest
from sqlalchemy import create_engine, text

from sqlalchemy.orm import Session


@pytest.fixture(scope="function")
def db_session():
    engine = create_engine("postgresql://scott:tiger@localhost/test")
    # connect to the database
    connection = engine.connect()

    # begin a non-ORM transaction
    trans = connection.begin()

    # bind an individual Session to the connection, selecting
    # "create_savepoint" join_transaction_mode
    session = Session(
            bind=connection,
            join_transaction_mode="create_savepoint"
        )

    yield session

    session.close()

    # rollback - everything that happened with the
    # Session above (including calls to commit())
    # is rolled back.
    trans.rollback()

    # return connection to the Engine
    connection.close()

    # since you are using create_engine() inside the decorator, dispose it
    engine.dispose()


def test_function_with_begin(db_session):
    # Let's execute some sql before
    db_session.execute(text("select 1"))
    # Let's call the function with begin()
    function_with_begin(db_session)


def function_with_begin(sess):
    # We will catch the exception here
    with sess.begin():
        sess.execute(text("select 1"))


```
Run begin() after select causes an error. 
` sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.`
Can you explain why it happens and how properly it handle?

**Reply by CaselIT** [reply]
the execute before calling begin will automatically create a transaction on the session, meaning that you cannot call begin again without closing the previous transaction

**Reply by aramazanov** [reply]
Should I call db_session.close() every time in such situations? The first time I did so but I thought it looks weird and I'm doing something wrong. Are there some best practices?

**Reply by CaselIT** [reply, marked as answer]
commit or rollback are probably more appropriate in that situation.
